drivers/net/ethernet/intel/idpf/eth/idpf_eth.c

```diff
@@ -0,0 +1,380 @@
+	vport = adapter->vport;
+	if (vport && !test_bit(IDPF_ETH_RESET_IN_PROG, adapter->flags))
+		idpf_send_get_stats_msg(vport);
+
+	queue_delayed_work(adapter->stats_wq, &adapter->stats_task,
+			   msecs_to_jiffies(10000));
+}
+
+/**
+ * idpf_eth_adapter_alloc - Allocated ethernet adapter struct
+ * @pdev: PCI device id of primary device
+ */
+static struct idpf_eth_adapter *idpf_eth_adapter_alloc(struct pci_dev *pdev)
+{
+	struct idpf_eth_adapter *adapter;
+	struct device *dev = &pdev->dev;
+
+	adapter = kzalloc(sizeof(*adapter), GFP_KERNEL);
+	if (!adapter)
+		return NULL;
+
+	adapter->post_init_wq = alloc_workqueue("%s-%s-pinit", 0, 0,
+						dev_driver_string(dev),
+						dev_name(dev));
+	if (!adapter->post_init_wq) {
+		dev_err(dev, "Failed to allocate post init workqueue\n");
+		kfree(adapter);
+		return NULL;
+	}
+
+	adapter->stats_wq = alloc_workqueue("%s-%s-stats", 0, 0,
+					    dev_driver_string(dev),
+					    dev_name(dev));
+	if (!adapter->stats_wq) {
+		dev_err(dev, "Failed to allocate workqueue\n");
+		kfree(adapter);
+		goto err_stats_wq;
+	}
+
+	INIT_DELAYED_WORK(&adapter->post_init_task,
+			  idpf_eth_device_post_init_task);
+	INIT_DELAYED_WORK(&adapter->stats_task, idpf_eth_statistics_task);
+	mutex_init(&adapter->vport_ctrl_lock);
+
+	return adapter;
+
+err_stats_wq:
+	destroy_workqueue(adapter->post_init_wq);
+	kfree(adapter);
+	return NULL;
+}
+
+/**
+ * idpf_eth_device_post_init - Delayed ethernet post initialization task
+ * @adapter: Ethernet private data struct
+ */
+static void idpf_eth_device_post_init(struct idpf_eth_adapter *adapter)
+{
+	struct idpf_vport_config *vport_config;
+
+	vport_config = adapter->vport_config;
+	/* Vport/Netdev is ready for initialization */
+	if (adapter->netdev &&
+	    !test_bit(IDPF_VPORT_REG_NETDEV, vport_config->flags)) {
+		register_netdev(adapter->netdev);
+		set_bit(IDPF_VPORT_REG_NETDEV, vport_config->flags);
+	}
+
+	/* Start the statistics task now */
+	queue_delayed_work(adapter->stats_wq, &adapter->stats_task, 0);
+}
+
+/**
+ * idpf_eth_device_post_init_task - Delayed ethernet post initialization task
+ * @work: work_struct handle to ethernet adapter data
+ */
+static void idpf_eth_device_post_init_task(struct work_struct *work)
+{
+	struct idpf_eth_shared *eth_shared;
+	struct idpf_eth_adapter *adapter;
+	struct pci_dev *pdev;
+
+	adapter = container_of(work, struct idpf_eth_adapter,
+			       post_init_task.work);
+	eth_shared = idpf_eth_adapter_shared(adapter);
+	pdev = eth_shared->pdev;
+
+	/* Continue to schedule work if some probe pre-init(s) are pending */
+	if (!adapter->dev_info->start_probe_post_init) {
+		queue_delayed_work(adapter->post_init_wq,
+				   &adapter->post_init_task,
+				   msecs_to_jiffies(1 * (pdev->devfn & 0x07)));
+		return;
+	}
+
+	/* Ready to finish the post init now */
+	idpf_eth_device_post_init(adapter);
+}
+
+/**
+ * idpf_eth_device_pre_init - Ethernet device pre-initialization routine
```

Comment:
If possible avoid the pre-init and post-init terminology. If there is such a thing as pre-init, it would be in main driver and as part of the "pre-init" it is allocating aux device, core eth dev info and initializing these 2 structures. Thats how I would look at "pre-init". So there is nothing like post-init - its just probe.  Its possible though that I have not understood the rationale. So please elaborate.

Reply:
It is pre-init and post init portion of the same probe under auxiliary device. This whole code belongs to the auxiliary driver, so it shouldn't create any confusion.